src/plugins/dnf/product_id.py

```diff
@@ -165,30 +163,45 @@ def _get_available():
         return available
 
     @staticmethod
-    def __write_cache_file(data, file_name):
+    def _write_cache_file(data, file_name):
         try:
             dir_name = os.path.dirname(file_name)
-            if not os.access(dir_name, os.R_OK):
+            cache_dir_exists: bool = False
+            if os.access(dir_name, os.R_OK):
+                cache_dir_exists = True
+            else:
                 log.debug("Try to create directory: %s" % dir_name)
-                os.makedirs(dir_name)
-            with open(file_name, "w") as file:
-                json.dump(data, file, default=json.encode)
-            log.debug("Wrote cache: %s" % file_name)
+                try:
+                    os.makedirs(dir_name, exist_ok=True)
+                except PermissionError as err:
+                    log.warning(f"Unable to create directory {dir_name}: {err}")
+                else:
+                    cache_dir_exists = True
+            if cache_dir_exists is True:
+                try:
+                    with open(file_name, "w") as file:
+                        json.dump(data, file, default=json.encode)
+                except PermissionError as err:
+                    log.warning(f"Unable to write to cache file {file_name}: {err}")
+                else:
+                    log.debug("Wrote cache: %s" % file_name)
         except IOError as err:
             log.error("Unable to write cache: %s" % file_name)
             log.exception(err)
```

Comment:
is this "global" IOError catch for the function still needed now, with the more specific exception handlers for directory creation and file opening? maybe it would be worth to fold it as additional handler for the try...except block that writes the json file

Reply:
Yeah, I also do not like the big block of code inside try-except. I will try to improve it.